src/tablestore.rs

```diff
@@ -210,6 +212,80 @@ impl TableStore {
         }
     }
 
+    #[allow(dead_code)]
+    pub(crate) async fn delete_sst(&self, id: &SsTableId) -> Result<(), SlateDBError> {
+        let path = self.path(id);
+        self.object_store
+            .delete(&path)
+            .await
+            .map_err(SlateDBError::ObjectStoreError)
+    }
+
+    pub(crate) async fn collect_garbage_wal(
+        &self,
+        min_age: Duration,
+        state: &CoreDbState,
+    ) -> Result<(), SlateDBError> {
+        let min_age = chrono::Duration::from_std(min_age).expect("invalid duration");
+        let wal_list = self
+            // Delete all SSTs up to (but not including) the last compacted WAL SST
+            .get_wal_sst_list(..state.last_compacted_wal_sst_id)
+            .await?;
+        for wal_id in wal_list {
+            let path = self.path(&SsTableId::Wal(wal_id));
+            let metadata = self.object_store.head(&path).await?;
+            if metadata
+                .last_modified
+                .signed_duration_since(chrono::Utc::now())
+                > min_age
+            {
+                self.object_store.delete(&path).await?;
+            }
+        }
+        Ok(())
+    }
+
+    pub(crate) async fn collect_garbage_ssts(
+        &self,
+        min_age: Duration,
+        state: &CoreDbState,
+    ) -> Result<(), SlateDBError> {
+        let min_age = chrono::Duration::from_std(min_age).expect("invalid duration");
+
+        // TODO Exclude snapshot SSTs from GC once snapshots are implemented
+        // Get all active SSTs
+        let l0_sst_ids: Vec<SsTableId> = state.l0.iter().map(|h| h.id).collect();
```

Comment:
I think it would be cleaner for the GC process to do the reference checking and then just pass in a list of SSTs to retain. Generally I feel like we should try to avoid making tablestore aware of the db metadata. Plus this is going to get fairly complex when we have to deal with snapshots, which themselves also have an expiry mechanism. I don't think it makes sense to jam all that logic into tablestore.

Reply:
How about adding this logic in DbState? It knows the active set already.
Garbage collector has StoredManifest, which has DbState. It uses tables store to list all the files, passes all of them and gets the garbage list from DbState. What is valid vs not is internal to DbState.
On top of the returned list, garbage then applies the age filter.

TableStore has the list methods alone, for manifest, wal, compacted and l0.

When snapshots are implemented, there are multiple active manifests.
Perhaps, StoredManifest exposes the `get garbage list` method (it calls DbState's method). The logic of an item can be deleted if it is in all the garbage list for all manifests.

Reply:
I'm going to take a crack at Rohan's approach.

@vigneshc I don't think the DbState provides much value in the approach you're talking about. If we give it a list of all SSTs and it returns which are inactive, that's not much different from the GC getting the active SSTs (which it can get from DbState) and removing them from the list. I kind of like keeping `DbState` as dumb as possible

I also don't know exactly how snapshots will look when implemented. Centralizing the logic in GC seems like it'd give us more options depending on where the snapshot stuff lands (inside vs. outside DbState).